File: Python/ecc.py
##############################################################################
# ENSEMBLE OF CLASSIFIER CHAINS                                              #
# Copyright (C) 2025                                                         #
#                                                                            #
# This code is free software: you can redistribute it and/or modify it under #
# the terms of the GNU General Public License as published by the Free       #
# Software Foundation, either version 3 of the License, or (at your option)  #
# any later version. This code is distributed in the hope that it will be    #
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of     #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General   #
# Public License for more details.                                           #
#                                                                            #
# Prof. Elaine Cecilia Gatto - UFLA - Lavras, Minas Gerais, Brazil           #
# Prof. Ricardo Cerri - USP - São Carlos, São Paulo, Brazil                  #
# Prof. Mauri Ferrandin - UFSC - Blumenau, Santa Catarina, Brazil            #
# Prof. Celine Vens - Ku Leuven - Kortrijik, West Flanders, Belgium          #
# PhD Felipe Nakano Kenji - Ku Leuven - Kortrijik, West Flanders, Belgium    #
#                                                                            #
# BIOMAL - http://www.biomal.ufscar.br                                       #
#                                                                            #
##############################################################################

import sys
import platform
import os

#FolderRoot = os.path.expanduser('/lapix/arquivos/elaine/HPML.ECC/Python')
#os.chdir(FolderRoot)
#current_directory = os.getcwd()
#sys.path.append('..')

import time
import io
import pickle
import numpy as np
import pandas as pd
from sklearn.base import clone
from sklearn.multioutput import ClassifierChain
import os

class ECC:
    """
    Ensemble of Classifier Chains (ECC) for multilabel classification,
    with timing, memory size tracking, and prediction export functionality.
    """

    def __init__(self, model, n_chains=10):             
        """
        Initialize the ECC ensemble.

        Purpose:
        --------
        Create an ensemble of classifier chains with a given base model and number of chains.

        Parameters:
        -----------
        model : scikit-learn compatible classifier
            The base classifier to be cloned in each chain.
        n_chains : int, default=10
            The number of classifier chains in the ensemble.

        Returns:
        --------
        None

        Example usage:
        --------------
        >>> from sklearn.ensemble import RandomForestClassifier
        >>> model = RandomForestClassifier()
        >>> ecc = ECC(model=model, n_chains=5)
        """
        self.model = model
        self.n_chains = n_chains
        self.chains = []                   # List of ClassifierChain objects
        self.chain_train_times = []        # Training time per chain
        self.train_time_total = 0          # Total training time
        self.test_time_total = 0           # Total prediction time (sum over all chains)
        self.chain_model_sizes = []        # Size of each chain in bytes
        self.total_model_size = 0          # Sum of all model sizes


    def fit(self, X, Y):
        """
        Train the ECC ensemble on the input features and labels.

        Purpose:
        --------
        Fit multiple classifier chains on the training data, each with a random chain order,
        and measure the training time for each chain.

        Parameters:
        -----------
        X : array-like or pandas DataFrame
            Feature matrix for training.
        Y : array-like or pandas DataFrame
            Multi-label target matrix for training.

        Returns:
        --------
        None
            The method trains the ensemble chains in-place and records training times.

        Example usage:
        --------------
        >>> ecc = ECC(model, n_chains=10)
        >>> ecc.fit(X_train, Y_train)
        >>> print(ecc.train_time_total)       # total training time
        >>> print(ecc.chain_train_times)      # list of training times per chain
        """
        self.chains = []
        self.chain_train_times = []
        self.total_times = [0] * self.n_chains  # used also for prediction time summing
        self.train_time_total = 0

        for i in range(self.n_chains):
            chain = ClassifierChain(clone(self.model), order="random")
            start_time = time.time()
            chain.fit(X, Y)
            end_time = time.time()

            fit_time = end_time - start_time
            self.chains.append(chain)
            self.chain_train_times.append(fit_time)
            self.total_times[i] += fit_time

        self.train_time_total = sum(self.chain_train_times)




    def predict_proba(self, x):
        """
        Predict class probabilities for the input data.

        Objective:
            Computes the average predicted class probabilities from multiple chains/models.

        Parameters:
            x (array-like): Input data for which to predict class probabilities.

        Returns:
            np.ndarray: Array of predicted class probabilities averaged over all chains.

        Example:
            >>> model = YourModel()
            >>> model.fit(X_train, y_train)
            >>> probs = model.predict_proba(X_test)
        """
        if self.chains is None:
            raise Exception('Oh no no no no!', 'Model has not been fitted yet.')

        return np.array([chain.predict_proba(x) for chain in self.chains]).mean(axis=0)


    

    def predict(self, X, threshold=0.5):
        """
        Make binary predictions based on a fixed probability threshold.

        Purpose:
        --------
        Generate binary multilabel predictions by applying a fixed threshold to the predicted probabilities.

        Parameters:
        -----------
        X : array-like or DataFrame
            Feature matrix for prediction.
        threshold : float, optional (default=0.5)
            Probability threshold used to convert probabilities into binary predictions.

        Returns:
        --------
        ndarray
            Binary predictions for each label, where probabilities equal or above the threshold are assigned 1, else 0.

        Example usage:
        --------------
        >>> ecc = ECC(model, n_chains=10)
        >>> ecc.fit(X_train, Y_train)
        >>> binary_predictions = ecc.predict(X_test, threshold=0.5)
        """
        probas = self.predict_proba(X)
        return (probas >= threshold).astype(int)

    

    def predict_cardinality(self, X, Y_train):
        """
        Make binary predictions using a threshold derived from the label cardinality in the training set.

        Purpose:
        --------
        Generate binary multilabel predictions by applying a threshold based on the average label cardinality 
        (average number of active labels per instance) calculated from the training labels.

        Parameters:
        -----------
        X : array-like or DataFrame
            Feature matrix for prediction.
        Y_train : array-like or DataFrame
            Multi-label target matrix used to compute the label cardinality threshold.

        Returns:
        --------
        ndarray
            Binary predictions for each label, thresholded according to the training label cardinality.

        Example usage:
        --------------
        >>> ecc = ECC(model, n_chains=10)
        >>> ecc.fit(X_train, Y_train)
        >>> predictions = ecc.predict_cardinality(X_test, Y_train)
        """
        cardinality = Y_train.sum(axis=1).mean() / Y_train.shape[1]
        probas = self.predict_proba(X)
        return (probas >= cardinality).astype(int)

    

    def get_model_sizes(self):
        """
        Measure the memory size (in bytes) of each trained chain and the total ensemble.

        Purpose:
        --------
        Calculate and return the memory size occupied by each chain classifier as well as the total size
        of the ensemble, for analyzing the model's memory consumption.

        Parameters:
        -----------
        None besides self.

        Returns:
        --------
        tuple (list, int)
            - A list with the size in bytes of each chain (list of integers).
            - The total size in bytes summing all chains (integer).

        Example usage:
        --------------
        >>> ecc = ECC(model, n_chains=5)
        >>> ecc.fit(X_train, Y_train)
        >>> sizes_per_chain, total_size = ecc.get_model_sizes()
        >>> print(f"Sizes per chain: {sizes_per_chain}")
        >>> print(f"Total ensemble size: {total_size} bytes")
        """
        if not self.chains:
            raise Exception('Model has not been fitted yet.')

        sizes = []
        for chain in self.chains:
            buffer = io.BytesIO()
            pickle.dump(chain, buffer)
            size = buffer.tell()
            sizes.append(size)

        self.chain_model_sizes = sizes           # Atualiza aqui!
        self.total_model_size = sum(sizes)
        return self.chain_model_sizes, self.total_model_size  
        

File: R/ecc.R
cat("\n#########################################################")
cat("\n# HPML.ECC: START                                       #")
cat("\n#########################################################\n\n")

# clean
rm(list=ls())


##############################################################################
# ENSEMBLE OF CLASSIFIER CHAINS                                              #
# Copyright (C) 2025                                                         #
#                                                                            #
# This code is free software: you can redistribute it and/or modify it under #
# the terms of the GNU General Public License as published by the Free       #
# Software Foundation, either version 3 of the License, or (at your option)  #
# any later version. This code is distributed in the hope that it will be    #
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of     #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General   #
# Public License for more details.                                           #
#                                                                            #
# 1 - Prof PhD Elaine Cecilia Gatto                                          #
# 2 - Prof PhD Ricardo Cerri                                                 #
# 3 - Prof PhD Mauri Ferrandin                                               #
# 4 - Prof PhD Celine Vens                                                   #
# 5 - PhD Felipe Nakano Kenji                                                #
# 6 - Prof PhD Jesse Read                                                    #
#                                                                            #
# 1 = Federal University of São Carlos - UFSCar - https://www2.ufscar.br     #
# Campus São Carlos | Computer Department - DC - https://site.dc.ufscar.br | #
# Post Graduate Program in Computer Science - PPGCC                          # 
# http://ppgcc.dc.ufscar.br | Bioinformatics and Machine Learning Group      #
# BIOMAL - http://www.biomal.ufscar.br                                       # 
#                                                                            # 
# 1 = Federal University of Lavras - UFLA                                    #
#                                                                            # 
# 2 = State University of São Paulo - USP                                    #
#                                                                            # 
# 3 - Federal University of Santa Catarina Campus Blumenau - UFSC            #
# https://ufsc.br/                                                           #
#                                                                            #
# 4 and 5 - Katholieke Universiteit Leuven Campus Kulak Kortrijk Belgium     #
# Medicine Department - https://kulak.kuleuven.be/                           #
# https://kulak.kuleuven.be/nl/over_kulak/faculteiten/geneeskunde            #
#                                                                            #
# 6 - Ecole Polytechnique | Institut Polytechnique de Paris | 1 rue Honoré   #
# d’Estienne d’Orves - 91120 - Palaiseau - FRANCE                            #
#                                                                            #
##############################################################################

#getwd()



cat("\n##############################")
cat("\n# Set Work Space             #")
cat("\n##############################\n\n")
library(here)
library(stringr)
FolderRoot <- here::here()



cat("\n########################################")
cat("\n# R Options Configuration              #")
cat("\n########################################\n\n")
options(java.parameters = "-Xmx64g")  # JAVA
options(show.error.messages = TRUE)   # ERROR MESSAGES
options(scipen=20)                    # number of places after the comma
options(repos = c(CRAN = "https://cloud.r-project.org"))



cat("\n########################################")
cat("\n# Creating parameters list             #")
cat("\n########################################\n\n")
parameters = list()


cat("\n########################################")
cat("\n# Reading Datasets-Original.csv        #")
cat("\n########################################\n\n")
setwd(FolderRoot)
datasets <- data.frame(read.csv("datasets-original.csv"))
parameters$Datasets.List = datasets


cat("\n#####################################")
cat("\n# GET ARGUMENTS FROM COMMAND LINE   #")
cat("\n#####################################\n\n")
args <- commandArgs(TRUE)

config_file <- args[1]


# config_file = "~/HPML.ECC/config-files/ecc-GnegativeGO-1.csv"


parameters$Config.File$Name = config_file
if(file.exists(config_file)==FALSE){
  cat("\n################################################################")
  cat("\n# Missing Config File! Verify the following path:              #")
  cat("\n################################################################")
  cat("\n --------------->", config_file)
  break
} else {
  cat("\n########################################")
  cat("\n# Properly loaded configuration file!  #")
  cat("\n########################################\n\n")
}


cat("\n########################################")
cat("\n# Config File                          #")
cat("\n########################################\n\n")
config = data.frame(read.csv(config_file))
print(config)




cat("\n##################################################")
cat("\n# HPML.ECC: Getting Parameters                   #")
cat("\n##################################################\n\n")
FolderScripts = toString(config$Value[1])
FolderScripts = str_remove(FolderScripts, pattern = " ")
parameters$Config.File$FolderScripts = FolderScripts

dataset_path = toString(config$Value[2])
dataset_path = str_remove(dataset_path, pattern = " ")
parameters$Config.File$Dataset.Path = dataset_path

folderResults = toString(config$Value[3]) 
folderResults = str_remove(folderResults, pattern = " ")
parameters$Config.File$Folder.Results = folderResults

implementation = toString(config$Value[4])
implementation = str_remove(implementation, pattern = " ")
parameters$Config.File$Implementation = implementation

dataset_name = toString(config$Value[5])
dataset_name = str_remove(dataset_name, pattern = " ")
parameters$Config.File$Dataset.Name = dataset_name

number_dataset = as.numeric(config$Value[6])
parameters$Config.File$Number.Dataset = number_dataset

number_folds = as.numeric(config$Value[7])
parameters$Config.File$Number.Folds = number_folds

number_cores = as.numeric(config$Value[8])
parameters$Config.File$Number.Cores = number_cores

ds = datasets[number_dataset,]
parameters$Dataset.Info = ds

print(ds)


cat("\n########################################")
cat("\n# Loading R Sources                    #")
cat("\n########################################\n\n")
source(file.path(FolderScripts, "libraries.R"))
source(file.path(FolderScripts, "utils.R"))


cat("\n#########################################")
cat("\n# HPML.ECC: Get directories             #")
cat("\n#########################################\n\n")
diretorios <- directories(parameters)
parameters$Directories = diretorios

FolderE = paste(parameters$Directories$FolderResults, 
                "/ECC2", sep="")
if(dir.exists(FolderE)==FALSE){dir.create(FolderE)}
parameters$Directories$FolderECC2 = FolderE


cat("\n########################################")
cat("\n# Creating temporary processing folder #")
cat("\n########################################\n\n")
if (dir.exists(folderResults) == FALSE) {dir.create(folderResults)}



cat("\n####################################################################")
cat("\n# Checking the dataset tar.gz file                                 #")
cat("\n####################################################################\n\n")
str00 = paste(dataset_path, "/", ds$Name,".tar.gz", sep = "")
str00 = str_remove(str00, pattern = " ")

if(file.exists(str00)==FALSE){
  
  cat("\n######################################################################")
  cat("\n# The tar.gz file for the dataset to be processed does not exist!    #")
  cat("\n# Please pass the path of the tar.gz file in the configuration file! #")
  cat("\n# The path entered was: ", str00, "                                  #")
  cat("\n######################################################################\n\n")
  break
  
} else {
  
  cat("\n####################################################################")
  cat("\n# tar.gz file of the DATASET loaded correctly!                     #")
  cat("\n####################################################################\n\n")
  
  # COPIANDO
  str01 = paste("cp ", str00, " ", diretorios$folderDataset, sep = "")
  res = system(str01)
  if (res != 0) {
    cat("\nError: ", str01)
    break
  }
  
  # DESCOMPACTANDO
  str02 = paste("tar xzf ", diretorios$folderDataset, "/", ds$Name,
                ".tar.gz -C ", diretorios$folderDataset, sep = "")
  res = system(str02)
  if (res != 0) {
    cat("\nError: ", str02)
    break
  }
  
  #APAGANDO
  str03 = paste("rm ", diretorios$folderDataset, "/", ds$Name,
                ".tar.gz", sep = "")
  res = system(str03)
  if (res != 0) {
    cat("\nError: ", str03)
    break
  }
  
}




##############################################################################

if(implementation=="utiml"){
  # 
  # setwd(FolderScripts)
  # source("run-utiml.R")
  # 
  # cat("\n\n############################################################")
  #   cat("\n# RSCRIPT ECC START                                     #")
  #   cat("\n############################################################\n\n")
  # timeFinal <- system.time(results <- run.ecc.utiml(ds, 
  #                                                   dataset_name,
  #                                                   number_dataset, 
  #                                                   number_cores, 
  #                                                   number_folds, 
  #                                                   folderResults))  
  # 
  # cat("\n\n#####################################################")
  #   cat("\n# RSCRIPT SAVE RUNTIME                              #")
  #   cat("\n#####################################################\n\n")
  # result_set <- t(data.matrix(timeFinal))
  # setwd(diretorios$folderECC)
  # write.csv(result_set, "Runtime-Final.csv")
  # x.minutos =(1 * as.numeric(result_set[3]))/60
  # setwd(diretorios$folderECC)
  # write(x.minutos, "minutos.txt")
  # 
  # 
  # cat("\n\n#####################################################")
  #   cat("\n# RSCRIPT DELETE                                   #")
  #   cat("\n####################################################\n\n")
  # str5 = paste("rm -r ", diretorios$folderDataset, sep="")
  # print(system(str5))
  # 
  # 
  # 
  # cat("\n\n######################################################")
  #   cat("\n# RSCRIPT COPY TO GOOGLE DRIVE                       #")
  #   cat("\n######################################################\n\n")
  # origem = diretorios$folderECC
  # destino = paste("nuvem:ECC/Utiml/", dataset_name, sep="")
  # comando = paste("rclone -P copy ", origem, " ", destino, sep="")
  # cat("\n", comando, "\n") 
  # a = print(system(comando))
  # a = as.numeric(a)
  # if(a != 0) {
  #   stop("Erro RCLONE")
  #   quit("yes")
  # }
  # 
  
} else if(implementation=="rf"){
  
  source(file.path(FolderScripts, "run-python.R"))
  
  cat("\n#########################################################")
  cat("\n# HPML.ECC: ECC START                                   #")
  cat("\n#########################################################\n\n")
  timeFinal <- system.time(results <- run.ecc.python(parameters))  
  
  
  cat("\n#####################################################")
  cat("\n# HPML.ECC: SAVE RUNTIME                            #")
  cat("\n#####################################################\n\n")
  result_set <- t(data.matrix(timeFinal))
  setwd(diretorios$folderECC)
  write.csv(result_set, "runtime-script.csv")
  
  
  cat("\n###################################################")
  cat("\n# HPML.ECC: DELETE ALL TEMPORARY FILES            #")
  cat("\n###################################################\n\n")
  str5 = paste("rm -r ", diretorios$folderDataset, sep="")
  print(system(str5))
  
  # /tmp/ecc-GnegativeGO/ECC2
  str5 = paste("rm -r ", diretorios$folderResults, "/ECC2", sep="")
  print(system(str5))
  
  
  cat("\n##############################################################")
  cat("\n# HPML.ECC: COMPRESS RESULTS                                 #")
  cat("\n##############################################################\n\n")
  # base_dir   <- "/tmp/d-GnegativeGO/Tested"
  base_dir   <- parameters$Directories$folderECC
  
  # output_tar <- "/tmp/d-GnegativeGO/GnegativeGO-results-hpml.tar.gz"
  output_tar <- paste0(parameters$Directories$FolderResults, "/",
                       parameters$Dataset.Info$Name, "-results-ecc.tar.gz")
  
  cmd <- paste("tar -zcvf", output_tar, "-C", base_dir, ".")
  res = system(cmd)
  
  if (res != 0) {
    #system(paste("rm -r", base_dir))
    #print(res)
    message("\n\n Something went wrong in compressing results files \n\n")
  } else {
    cat("\n✅ Compressão concluída com sucesso!\n")
  }
  
  cat("\n##########################################################")
  cat("\n# ECC: COPY TO HOME                                      #")
  cat("\n##########################################################\n\n")
  str0 = paste0(FolderRoot, "/Reports")
  if(dir.exists(str0)==FALSE){dir.create(str0)}
  
  str4 <- paste0(parameters$Directories$folderResults, "/",
    parameters$Dataset.Info$Name, "-results-ecc.tar.gz")
  
  str5 = paste("cp ", str4, " ", str0, sep="")
  res = system(str5)
  
  if(res!=0){
    #system(paste("rm -r ", parameters$Directories$FolderResults, sep=""))
    #print(res)
    message("\n\n Something went wrong in compressing results files \n\n")
  }
  
  
} else {
  
  # 
  # setwd(FolderScripts)
  # source("run-mulan.R")
  # 
  # 
  # cat("\n\n############################################################")
  # cat("\n# RSCRIPT ECC START                                     #")
  # cat("\n############################################################\n\n")
  # timeFinal <- system.time(results <- run.ecc.mulan(ds, 
  #                                                   dataset_name,
  #                                                   number_dataset, 
  #                                                   number_cores, 
  #                                                   number_folds, 
  #                                                   folderResults))  
  # 
  # 
  # 
  # cat("\n\n#####################################################")
  # cat("\n# RSCRIPT SAVE RUNTIME                              #")
  # cat("\n#####################################################\n\n")
  # result_set <- t(data.matrix(timeFinal))
  # setwd(diretorios$folderECC)
  # write.csv(result_set, "Runtime-Final.csv")
  # x.minutos = (1 * as.numeric(result_set[3]))/60
  # setwd(diretorios$folderECC)
  # write(x.minutos, "minutos.txt")
  # 
  # 
  # cat("\n\n#################################################")
  # cat("\n# RSCRIPT DELETE                                  #")
  # cat("\n###################################################\n\n")
  # str5 = paste("rm -r ", diretorios$folderDataset, sep="")
  # print(system(str5))
  # 
  # 
  # cat("\n\n######################################################")
  #   cat("\n# RSCRIPT COPY TO GOOGLE DRIVE                       #")
  #   cat("\n######################################################\n\n")
  # origem = diretorios$folderECC
  # destino = paste("nuvem:ECC/Mulan/", dataset_name, sep="")
  # comando = paste("rclone -P copy ", origem, " ", destino, sep="")
  # cat("\n", comando, "\n") 
  # a = print(system(comando))
  # a = as.numeric(a)
  # if(a != 0) {
  #   stop("Erro RCLONE")
  #   quit("yes")
  # }
  # 
  
  
} 


# cat("\n\n###################################################################")
# cat("\n# ECC ECC: COMPRESS RESULTS                                      #")
# cat("\n#####################################################################\n\n")
# str3 = paste("tar -zcvf ", diretorios$folderResults, "/", 
#              dataset_name, "-results-ECC.tar.gz ", 
#              diretorios$folderResults, sep="")
# print(system(str3))


# cat("\n\n##############################################################")
# cat("\n# ECC ECC: COPY TO FOLDER REPORTS                           #")
# cat("\n###############################################################\n\n")
# str0 = "~/Ensemble-Classifier-Chains/Reports/"
# if(dir.exists(str0)==FALSE){dir.create(str0)}
# str1 = paste(diretorios$folderResults, "/", dataset_name,
#              "-results-ECC.tar.gz", sep="")
# str4 = paste("cp -r ", str1 , " ", str0, sep="")
# print(system(str4))



cat("\n#######################################################")
cat("\n# CLEAN                                               #")
cat("\n#######################################################\n\n")
cat("\nDelete folder \n")
str5 = paste("rm -r ", folderResults, sep="")
print(system(str5))


cat("\n################################################################")
cat("\n# RSCRIPT SUCCESSFULLY FINISHED                                #")
cat("\n################################################################\n\n")


rm(list = ls())
gc()

###############################################################################
# Please, any errors, contact us: elainececiliagatto@gmail.com                #
# Thank you very much!                                                        #                                #
###############################################################################
